### PR TITLE
Add lightning segmentation module + use Lightning CLI

### PR DESCRIPTION
**Pull request by LBerth**

### CODE REVIEW

File: mfai/torch/segmentation_module.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        df.to_csv(path_csv, index=False)
+        print(f"--> Metrics for all samples saved in \033[91m\033[1m{path_csv}\033[0m")
+
+    def on_test_epoch_end(self):
+        """Logs metrics in tensorboard hparams view, at the end of run."""
+        df = self.build_metrics_dataframe()
+        self.save_test_metrics_as_csv(df)
+        df = df.drop("Name", axis=1)
+
+    def last_activation(self, y_hat):
```
Comment (tourniert) [inline mfai/torch/segmentation_module.py:217]:
```suggestion
    def last_activation(self, y_hat: torch.Tensor):
```

File: mfai/torch/segmentation_module.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        """Applies appropriate activation according to task."""
+        if self.type_segmentation == "multiclass":
+            y_hat = y_hat.log_softmax(dim=1).exp()
+        elif self.type_segmentation in ["binary", "multilabel"]:
+            y_hat = torch.nn.functional.logsigmoid(y_hat).exp()
+        return y_hat
+
+    def probabilities_to_classes(self, y_hat):
```
Comment (tourniert) [inline mfai/torch/segmentation_module.py:225]:
```suggestion
    def probabilities_to_classes(self, y_hat: torch.Tensor):
```

**Review by tourniert — APPROVED** [review]
Good job !

File: mfai/torch/segmentation_module.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+import torchmetrics as tm
+import pandas as pd
+
+from typing import Literal, Callable
+from pathlib import Path
+from mfai.torch.models.base import ModelABC
+
+from pytorch_lightning.utilities import rank_zero_only
+
+# define custom scalar in tensorboard, to have 2 lines on same graph
```
[the remark this answers is not in the document]
Comment (LBerth) [reply mfai/torch/segmentation_module.py:12]: this one is specific to tensorboard, so I prefer to leave it

File: mfai/torch/segmentation_module.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+            y = y.to(memory_format=torch.channels_last)
+        # We prefer when the last activation function is included in the loss and not in the model.
+        # Consequently, we need to apply the last activation manually here, to get the real output.
+        y_hat = self.model(x)
+        loss = self.loss(y_hat, y)
+        y_hat = self.last_activation(y_hat)
+        return y_hat, loss
+
+    def on_train_start(self):
+        """Setup custom scalars panel on tensorboard and log hparams.
```
[the remark this answers is not in the document]
Comment (LBerth) [reply mfai/torch/segmentation_module.py:118]: idem

File: README.md [hunk 5]
```diff
@@ -67,6 +68,25 @@ Currently we support the following neural network architectures:
...
+The module provides:
+- customization for each stage of the training
+- metrics and plot logging with tensorboard
+- logging of configuration and hyperparameters
+- computation of several metrics during validation stage
+- test stage: compute metrics for each sample individualy and save them in CSV file
+
+# Lightning CLI
+
+Lightning CLI is a convinient way to easily configure your DL experiments and reduce the number of lines of code in your project.
```
Comment (colon3ltocard) [inline README.md:86]: add a link to the official doc ?

**Review by tourniert — APPROVED** [review]

File: README.md [hunk 6]
```diff
@@ -161,6 +181,250 @@ output_tensor = onnx_load_and_infer("model.onnx", input_tensor)
...
+
+The lightning module can be instantiated and used in a forward pass as follows:
+
+```python
+arch = UNet(in_channels=1, out_channels=1, input_shape=[64, 64])
+loss = torch.nn.MSELoss()
+model = SegmentationLightningModule(arch, "binary", loss)
+x = torch.randn((1, 1, 64, 64)).float()
+model(x)
+```
```
Comment (colon3ltocard) [inline README.md:195]: add import to have a self sufficient example ?

File: mfai/torch/dummy_dataset.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
+import torch
+from lightning.pytorch.core import LightningDataModule
+from torch.utils.data import DataLoader, Dataset
+
+
+class DummyDataset(Dataset):
```
Comment (colon3ltocard) [inline mfai/torch/dummy_dataset.py:6]: top level succinct comment/doc ?

File: mfai/torch/dummy_dataset.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        x = torch.randn((self.nb_input_channels, self.dim_x, self.dim_y)).float()
+        y = torch.randint(
+            0, 1, (self.nb_output_channels, self.dim_x, self.dim_y)
+        ).float()
+        if self.split == "test":
+            return x, y, index  # return name of sample for test step
+        return x, y
+
+
+class DummyDataModule(LightningDataModule):
```
Comment (colon3ltocard) [inline mfai/torch/dummy_dataset.py:35]: top level succinct comment/doc ?

File: mfai/torch/models/base.py [hunk 9]
```diff
@@ -26,3 +26,9 @@ def input_spatial_dims(self) -> Tuple[int, ...]:
         A model supporting 2d tensors should return (2,).
         A model supporting 2d and 3d tensors should return (2, 3).
         """
+
+    def check_required_attributes(self):
```
Comment (colon3ltocard) [inline mfai/torch/models/base.py:30]: maybe a comment to explain why we do that ? naming is explicit but i never hurts

File: README.md [hunk 10]
```diff
@@ -67,6 +68,25 @@ Currently we support the following neural network architectures:
...
+The module provides:
+- customization for each stage of the training
+- metrics and plot logging with tensorboard
+- logging of configuration and hyperparameters
+- computation of several metrics during validation stage
+- test stage: compute metrics for each sample individualy and save them in CSV file
+
+# Lightning CLI
+
+[Lightning CLI](https://lightning.ai/docs/pytorch/stable/cli/lightning_cli.html#lightning-cli) is a convinient way to easily configure your DL experiments and reduce the number of lines of code in your project.
```
Comment (colon3ltocard) [inline README.md:86]:
```suggestion
[Lightning CLI](https://lightning.ai/docs/pytorch/stable/cli/lightning_cli.html#lightning-cli) is a convenient way to easily configure your DL experiments and reduce the number of lines of code in your project.
```

**Review by colon3ltocard — APPROVED** [review]
nice work !

**Review by tourniert — APPROVED** [review]

**Review by tourniert — APPROVED** [review]

**Review by colon3ltocard — APPROVED** [review]
